packages/flutter/lib/src/material/button_theme.dart

```diff
@@ -0,0 +1,263 @@
+    EdgeInsetsGeometry padding,
+    ShapeBorder shape,
+  }) : assert(textTheme != null),
+       assert(minWidth != null && minWidth >= 0.0),
+       assert(height != null && height >= 0.0),
+       _padding = padding,
+       _shape = shape;
+
+  /// The minimum width for buttons.
+  ///
+  /// The actual horizontal space allocated for a button's child is
+  /// at least this value less the theme's horizontal [padding].
+  ///
+  /// Defaults to 88.0 logical pixels.
+  final double minWidth;
+
+  /// The minimum height for buttons.
+  ///
+  /// The actual vertical space allocated for a button's child is
+  /// at least this value less the theme's horizontal [padding].
+  ///
+  /// Defaults to 36.0 logical pixels.
+  final double height;
+
+  /// Defines a button's base colors, and the defaults for the button's minimum
+  /// size, internal padding, and shape.
+  final ButtonTextTheme textTheme;
+
+  /// Simply a convenience that returns [minWidth] and [height] as a
+  /// [BoxConstraints] object:
+  /// ```dart
+  /// return new BoxConstraints(
+  ///   minWidth: minWidth,
+  ///    minHeight: height,
+  /// );
+  /// ```
+  BoxConstraints get constraints {
+    return new BoxConstraints(
+      minWidth: minWidth,
+      minHeight: height,
+    );
+  }
+
+  /// Padding for a button's child (typically the button's label).
+  ///
+  /// Defaults to 24.0 on the left and right if [textTheme] is
+  /// [ButtonTextTheme.primary], 16.0 on the left and right otherwise.
+  EdgeInsetsGeometry get padding {
+    if (_padding != null)
+      return _padding;
+    switch (textTheme) {
+      case ButtonTextTheme.normal:
+      case ButtonTextTheme.accent:
+        return const EdgeInsets.symmetric(horizontal: 16.0);
+      case ButtonTextTheme.primary:
+        return const EdgeInsets.symmetric(horizontal: 24.0);
+    }
+    return EdgeInsets.zero;
+  }
+  final EdgeInsetsGeometry _padding;
+
+  /// The shape of a button's material.
+  ///
+  /// The button's highlight and splash are clipped to this shape. If the
+  /// button has an elevation, then its drop shadow is defined by this
+  /// shape as well.
+  ///
+  /// Defaults to a rounded rectangle with circular corner radii of 4.0 if
+  /// [textTheme] is [ButtonTextTheme.primary], a rounded rectangle with
+  /// circular corner radii of 2.0 otherwise.
+  ShapeBorder get shape {
+    if (_shape != null)
+      return _shape;
+    switch (textTheme) {
+      case ButtonTextTheme.normal:
+      case ButtonTextTheme.accent:
+        return const RoundedRectangleBorder(
+          borderRadius: const BorderRadius.all(const Radius.circular(2.0)),
+        );
+      case ButtonTextTheme.primary:
+        return const RoundedRectangleBorder(
+          borderRadius: const BorderRadius.all(const Radius.circular(4.0)),
+        );
+    }
+    return const RoundedRectangleBorder();
+  }
+  final ShapeBorder _shape;
+
+  @override
+  bool operator ==(dynamic other) {
+    if (other is! ButtonThemeData)
```

Comment:
check the exact runtimeType rather than using `is`, so that subclassing works right